### ISSUE TITLE
Support for Windows

### ISSUE THREAD

**Issue by mhjlam** [davmac314/dasynq#4]
Hello!

I am searching for a more modern alternative to the available C-based event loop libraries, and dasynq looks very promising. I'm particularly impressed by the clean code and excellent documentation!

It is mentioned in the documentation that dasynq is cross-platform, and as such I was hoping that this library may be used on Windows as well. I am however noticing that there is a hard dependency on the unistd.h header, which is unavailable on the Windows platform. Digging a little deeper, the functions used from this header have mostly to do with file descriptors, for which WinAPI alternatives might possibly be employed as well.

Are there any plans to have dasynq support the Windows platform in the near future, or could some pointers be given of how to go about doing this manually?

Kind regards,
Maurits

**Comment by davmac314** [comment]
Ok! Let's tackle the easy part first:

> Are there any plans to have dasynq support the Windows platform in the near future

I kept Windows (and non-POSIX systems generally) in mind when writing Dasynq, but I have no plans to port to Windows at the moment. I'd certainly be willing to accept Windows support as a contribution (assuming code and documentation quality is maintained to the current standard).

Now:
> could some pointers be given of how to go about doing this manually

The part you really want to keep (but with some modification) is the `event_dispatch` class, and maybe the `event_loop` class (both defined in dasynq.h). The latter is the main Dasynq interface.

For `event_dispatch`:

The `event_dispatch` class is responsible for maintaing a queue of events to dispatch (in priority order). There is still some POSIX-specific code here (to submit certain even types) but this could be separated out, I think, without too much trouble. At its core `event_dispatch` does the following:

- maintains the queue of events
- allows submitting new events (`queue_watcher`, which the various `receive_*` functions call)
- allows retrieving events from the queue, to be dispatched (`pull_event`)
- allows issuing a delete for a watcher, which dequeues it immediately unless it is currently being processed (`issue_delete`). This has an overload with special handling for `base_bidi_fd_watcher` type.

The main changes needed would be to change the watcher-specific methods (`receive_*` functions). So, you either want a windows specific version of the class, or you want to factor out those platform-specific functions somehow (introduce a new platform-specific base class for example).

For `event_loop`:

This is the primary interface of the Dasynq library, but in terms of function, its main purpose is to abstract away details of the backend. As part of this, it has locking logic to prevent the backend from being polled by more than one thread at once. See eg `get_attn_lock`, `poll_attn_lock`, `get_pollwait_lock`, `release_lock`. There's a reasonable explanation of how these are meant to be used in the code comments (starting around line 715). (In fact that whole comment block is worth reading as it explains the essential problems that Dasynq has to solve).

These locking functions, assuming they are needed for a Windows backend, and for eg the `poll` and `run` functions, could be kept; most of the rest of the class is probably platform specific.

In general:

- I would suggest start by copying dasynq.h and removing all the platform-specific stuff, but keep (most of) the two classes outlined above
- implement watchers specific for Windows event types, derived from `base_watcher` which is defined in base_watchers.h. You'll probably need a Windows-specific version of this header as well, or separate out the platform-specific parts
- add functions to `event_loop` to register the different watcher types
- implement a Windows backend, see for example the epoll backend (epoll.h or dasynq-epoll.h). This is mostly about adding certain watch types, and pulling events. `event_loop` will call the `add_*` functions directly.

Possibly also:

- to make it truly cross-platform, you'd need abstractions for various things that are done differently on the different platforms. Eg opening a connection to a socket, and creating a watcher for it, etc.

In Summary:

It would be a lot of work, but should definitely be possible! You could start with a small number of watcher types (just one or two) and work from there. I would start by just duplicating code (copying headers, eg dasynq.h to dasynq-windows.h) and worry about factoring out common code later.

**Comment by mhjlam** [comment]
Hi Davin,

Thanks for the detailed response! I will have a closer look at the parts of the library you've pointed out. Unfortunately I cannot say for sure that I will get far enough to make a contribution, because as you said it would be a lot of work; I would need to learn the internals of the library first, then find accurate WinAPI replacements for POSIX calls, and then make sure that all the current functionality works similarly on a Windows platform. However, I will make an attempt. If nothing else, the above explanation is a useful guideline for anyone who wishes to attempt the same.

Thanks again!
